### ISSUE TITLE
rebin logic: the first/last out_edge needed?

### ISSUE THREAD

**Issue by chunhochow** [lbl-anp/becquerel#209]
@jccurtis :

Do you remember the logic for the rebinning? I'm trying to update it to use numba.guvectorize rather than numba.jit (because the jit isn't "vectorized enough" for xarray to interpret it as a vectorized function.

The issue with numba.guvectorize is that it requires the shape for the inputs and outputs, and so out_edges = (m), and out_spectrum = (m-1) (except that you can't specify the "-1". An easy way to deal with this is to just knock out the first/last edge in out_edges. I was trying to go through the logic of _rebin_interpolate , and it seems that we don't actually need the first/last edge, because we're putting all overflow counts in the first/last bin anyways. Do you remember differently?

**Comment by cosama** [comment]
Yeah, overflows right now seem to end up in the first and last bin. I have been dealing with this with:

```python
new_cnts = bq.rebin(cnts, new_edges, [-np.inf, *edges, np.inf])
overflow = (new_cnts[0], new_cnts[-1])
new_cnts = new_cnts[1:-1]
```

**Comment by chunhochow** [comment]
Gotcha, yes, having some option of that sort would make sense @cosama , for the future.

OK, so the first and last bin edges really seem useless then. I'll try to remove the use of the last bin edge in the code, which would allow me to vectorize it (because input dim = len(out_edges) needs to be equal to output dim = len(out_spectrum). And then we can think about adding that option.

I'm overhauling the code to allow the use of xarray and dask (to handle huge arrays in out-of-memory calculations).

**Comment by cosama** [comment]
I am a bit surprised why you compare edges to spectrum, we have for quantities:

 1) `input_spectrum` dimension `(N,)`
 2) `input_edges` dimension `(N + 1,)`
 3) `output_spectrum` dimension `(M, )`
 4) `output_edges` dimension `(M + 1`,)`

All of these dimensions are known when calling `bq.rebin`. 1., 2. and 4. are parameters for `bq.rebin`, 3. can be calculated from 4. If necessary we could add a `output` arg, that takes a `np.array` (or whatever) similar to many `numpy` functions, so that you don't have to create the output array in the function itself.

Does this make sense?

**Comment by jvavrek** [comment]
Do we ever reach huge arrays in `becquerel`? Even a 32k bin array is pretty trivial for `numpy`

**Comment by cosama** [comment]
I agree with @jvavrek  also, not sure when we have spectra that would not fit in memory. This is very different when considering multi-dimensional histograms. There is support for arrays of spectra at this point, I assume that is what @chunhochow is working on here.

**Comment by chunhochow** [comment]
@cosama exactly, the M+1 -> M is a problem because numba's guvectorize requires the output dimensions to be derivable from the input dimensions, and they don't seem to allow "M' - 1" as the dim. It HAS to be the dimensions (directly) defined in the inputs. Putting in an output as a input seems like it might also be investigatable, but numba doesn't seem to like side-effects (changing inputs); it has to be done super carefully.

@jvavrek yes we're having data that's like 15GB++ which is way too big for RAM. I'm also updating this to allow the use of dask and xarray, which allows for out of memory calculations and other parallelization benefits.

I'm just working on 2D arrays. This is for PANDA.

**Comment by chunhochow** [comment]
Right now how @markbandstra and @dhellfeld are dealing with it is to read data in in chunks, then rebinning each chunk before moving onto the next one. But if we can just (gu)vectorize the single spectrum/1D functions, then you can just directly run that over the entire 15GB, and let dask/xarray handle it for you natively. (instead of writing a ton of code just to handle reading things and processing things in chunks, which is a solved problem (in the outside world))

**Comment by jvavrek** [comment]
Ah yeah, 2D histograms will get you there. Are multi-dimensional histograms planned for `becquerel`?

**Comment by chunhochow** [comment]
haha i don't know about that. @cosama what do you mean about multi-dim (like what are the dims?) If yes then yes we should definitely integrate xarray in, which makes my life slightly easier (because I don't have to consider if I want to bring xarray in as a separate dependency)

**Comment by cosama** [comment]
No multi-dimensional support yet, as I said only thing we have is an array of spectra.

**Comment by jccurtis** [comment]
@jvavrek multi dimensional histograms are dealt with our internal tool `curie`. We don't have rebinning implemented yet. `bq` does currently allow rebining over 2D arrays (n spectra by energy). It just iterates in `numba` and calls the 1D rebin function.

@chunhochow this sounds like a cool application but I have to ask: since the rebin operations all independent would a map/apply with lots of cores/memory be a lower lift here? These are not matrix operations where you need a bunch of rows/columns at once.

**Comment by chunhochow** [comment]
@jccurtis exactly! but maps require vectorized operations to be quick, and the way to do it easily is using `numba.guvectorize`. (actually `nb.jit`, as we had implemented, is mostly there, but it's not "as" recognised as a vectorized function, and unfortunately it doesn't work with `xarray`

So what I'm doing is trying to guvectorize the 1D operations. The 2D operations then would be even cleaner than before because if I do a vectorized mapping I believe it can handle both the 1D and 2D case (so we won't need to have a separate 2D operation). 

See http://xarray.pydata.org/en/stable/generated/xarray.apply_ufunc.html

**Comment by chunhochow** [comment]
Wait, are y'all trying to partially-reimplement functions that you can find in `xarray` in `curie`.

**Comment by cosama** [comment]
I think this has been resolved with #211. Closing.
